packages/forms/src/UIForm/fieldsets/Array/Array.component.js

```diff
@@ -0,0 +1,192 @@
+		indexedKeys[firstIndexPlaceholder] = index;
+	}
+	return indexedKeys;
+}
+
+function getRange(previousIndex, nextIndex) {
+	if (previousIndex < nextIndex) {
+		return {
+			minIndex: previousIndex,
+			maxIndex: nextIndex + 1,
+		};
+	}
+
+	return {
+		minIndex: nextIndex,
+		maxIndex: previousIndex + 1,
+	};
+}
+
+function getItemSchema(arraySchema, index) {
+	// insert index in all fields
+	const items = arraySchema.items.map(item => ({
+		...item,
+		key: adaptKeyWithIndex(item.key, index),
+	}));
+
+	// insert index in item schema key
+	const key = arraySchema.key.concat(index);
+
+	return {
+		key,
+		items,
+		widget: arraySchema.itemWidget || 'fieldset',
+	};
+}
+
+export default class ArrayWidget extends React.Component {
+	constructor(props) {
+		super(props);
+
+		this.onAdd = this.onAdd.bind(this);
+		this.onRemove = this.onRemove.bind(this);
+		this.onReorder = this.onReorder.bind(this);
+	}
+
+	onAdd(event) {
+		const arrayMergedSchema = this.props.schema;
+		const defaultValue = arrayMergedSchema.schema.items.type === 'object' ? {} : '';
+		const value = [...this.props.value, defaultValue];
+
+		this.props.onChange(
+			event,
+			{ schema: arrayMergedSchema, value }
+		);
+	}
+
+	onRemove(event, indexToRemove) {
+		const schema = this.props.schema;
+		const value = this.props.value.slice(0);
+		value.splice(indexToRemove, 1);
+
+		return this.props.onChange(
+			event,
+			{ schema, value },
+			{
+				widgetChangeErrors(errors) {
+					// shift up the items errors after the one we remove
+					return shiftArrayErrorsKeys(
+						errors,
+						{
+							arrayKey: schema.key,
+							minIndex: indexToRemove,
+							shouldRemoveIndex(index) {
+								return index === indexToRemove;
+							},
+							getNextIndex(index) {
+								return index - 1;
+							},
+						},
+					);
+				},
+			}
+		);
+	}
+
+	onReorder(event, { previousIndex, nextIndex }) {
+		const schema = this.props.schema;
+		const value = this.props.value.slice(0);
+		const [item] = value.splice(previousIndex, 1);
+		value.splice(nextIndex, 0, item);
+
+		return this.props.onChange(
+			event,
+			{ schema, value },
+			{
+				widgetChangeErrors(errors) {
+					// determine the range [min, max[ of items to shift, with the pace
+					const { minIndex, maxIndex } = getRange(previousIndex, nextIndex);
+					const switchPace = Math.sign(previousIndex - nextIndex);
+
+					// shift the items errors between the previous and next position
+					// set the item-we-move errors indexes
+					return shiftArrayErrorsKeys(
+						errors,
+						{
+							arrayKey: schema.key,
+							minIndex,
+							maxIndex,
+							getNextIndex(index) {
+								return index === previousIndex ? nextIndex : index + switchPace;
+							},
+						},
+					);
+				},
+			}
+		);
+	}
+
+	render() {
+		const { errorMessage, id, isValid, schema, value, ...restProps } = this.props;
+
+		return (
+			<div>
+				<ol id={id} className={theme['tf-array']}>
+					{value.map((itemValue, index) => {
```

Comment:
I'm always a little reluctent to separate ol/li because they are tied :/
Here the code is not too big IMO, and we have a clear view of the markup without scrolling.